### ISSUE TITLE
Resizing datagrid can lead to an infinite loop

### ISSUE THREAD

**Issue by budul100** [objorke/PropertyTools#176]
The size update mechanism of the datagrid can lead to an infinite loop since the determination of the available width isn't considering the extra column for the scroll bar (see line 4159 in [DataGrid.cs](https://github.com/objorke/PropertyTools/blob/develop/Source/PropertyTools.Wpf/DataGrid/DataGrid.cs)

`            
            var availableWidth = this.sheetScrollViewer.ViewportWidth;

            var widthPerStar = Math.Max((availableWidth - usedWidth) / starsToDistribute, 0);
            for (var i = 0; i < this.Columns; i++)
            {
                var columnWidth = this.GetColumnWidth(i);
                if (columnWidth.IsStar)
                {
                    this.SetColumnWidth(i, new GridLength(widthPerStar * columnWidth.Value));
                }
            }
`

There is an extra column add which size is missing at the upper determination of the availableWidth (see line 4240 in [DataGrid.cs](https://github.com/objorke/PropertyTools/blob/develop/Source/PropertyTools.Wpf/DataGrid/DataGrid.cs).

` 
           // Add one empty column covering the vertical scrollbar
            this.columnGrid.ColumnDefinitions.Add(
                new System.Windows.Controls.ColumnDefinition { Width = new GridLength(40) });
` 

I have solved the problem by changing line 4159 into:

`
            var availableWidth = this.sheetScrollViewer.ViewportWidth - columnGrid.ColumnDefinitions.Last().Width.Value;
`

**Comment by objorke** [comment]
Thanks for the report and a possible solution! Can this be reproduced in one of the existing examples? Otherwise I need an example before we can change this. 

I see the implementation for the horizontal and vertical scrollbars are different:
https://github.com/objorke/PropertyTools/blob/develop/Source/PropertyTools.Wpf/DataGrid/DataGrid.cs#L4240
https://github.com/objorke/PropertyTools/blob/develop/Source/PropertyTools.Wpf/DataGrid/DataGrid.cs#L4478
Not sure why it was done like this...

**Comment by budul100** [comment]
You can reproduce this issue in the existing examples. If you replace the XAML of [Source/Examples/DataGrid/DataGridDemo/Examples/ListOfIntExample.xaml](https://github.com/objorke/PropertyTools/blob/develop/Source/Examples/DataGrid/DataGridDemo/Examples/ListOfIntExample.xaml) with the following code and open the List<int> example, then it will lead into the described infinite loop:

```
<Window x:Class="DataGridDemo.ListOfIntExample"
        xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:p="clr-namespace:PropertyTools.Wpf;assembly=PropertyTools.Wpf"
        Title="List&lt;int&gt;" Height="450" Width="300">
    <ScrollViewer 
        Height="Auto" 
        Width="Auto"
        HorizontalScrollBarVisibility="Auto"
        VerticalScrollBarVisibility="Auto">
        <p:DataGrid ItemsSource="{Binding ItemsSource}"/>
    </ScrollViewer>
</Window>
```

**Comment by objorke** [comment]
I have added the example (see ScrollViewerExample) but not applied the proposed change - on my machine it is not working correctly when reducing the size of the window.

**Comment by budul100** [comment]
Yes, I've recognized the same problem. My suggestion is more a kind of workaround for other users. 

If I found a better solution, I'll inform you.

**Comment by MD-73** [comment]
The example consists of a DataGrid inside a ScrollViewer. With @budul100's fix applied, the DataGrid does not resize when the window size is reduced. I think this is not too unexpected and this problem disappears if the outer container is changed from a ScrollViewer to say a Grid.

There is then a minor problem (depending on the intended behaviour) when vertically resizing the window: The grid area and scroll bar column "jitter" as their widths change with the visibility of the vertical scroll bar. This can be fixed by using ActualWidth rather than ViewPortWidth, i.e. change the [line](https://github.com/objorke/PropertyTools/blob/191efdf9de631d10409c424e289c3826a19a2bd6/Source/PropertyTools.Wpf/DataGrid/DataGrid.cs#L4091) instead to

```
var availableWidth = this.sheetScrollViewer.ActualWidth - columnGrid.ColumnDefinitions.Last().Width.Value;
```

**Comment by objorke** [comment]
Cannot observe the inifinite loop or jitter in the included example. Closing this as not reproducable, but please re-open if this is still an issue!
